### PR TITLE
Add new IcfgBuilder

### PR DESCRIPTION
**Pull request by Heizmann**
Commits for switching from RCFGBuilder to IcfgBuilder are missing in this branch. But Frank convinced me to already start a pull request.

### CODE REVIEW

File: trunk/examples/settings/automizer/ForwardPredicates_IcfgBuilder.epf [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#Thu Jan 30 13:20:28 CET 2014
```
Comment (schuessf) [inline trunk/examples/settings/automizer/ForwardPredicates_IcfgBuilder.epf]: These settings can be removed, once `IcfgBuilder` is the default.

File: trunk/examples/toolchains/AutomizerBpl_IcfgBuilder.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<rundefinition>
```
Comment (schuessf) [inline trunk/examples/toolchains/AutomizerBpl_IcfgBuilder.xml]: This toolchain can be removed, once `IcfgBuilder` is the default.

File: trunk/examples/Backtranslation/regression/bpl/Pointer.errorpath [hunk 3]
```diff
@@ -1,38 +1,40 @@
...
-       VAL        [#NULL!base=0, #NULL!offset=0, #t~mem3=4]
-[L18]  COND TRUE  #t~mem3 == 4
-       VAL        [#NULL!base=0, #NULL!offset=0]
-[L19]             havoc #t~mem3;
-       VAL        [#NULL!base=0, #NULL!offset=0]
-[L20]  FCALL      call assert_fail();
-       VAL        [#NULL!base=0, #NULL!offset=0]
-[L31]             assert false;
-       VAL        [#NULL!base=0, #NULL!offset=0]
+[L49]  CALL       call ULTIMATE.init();
```
Comment (schuessf) [inline trunk/examples/Backtranslation/regression/bpl/Pointer.errorpath]: What has changed in the backtranslation that is relevant for this test?

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/preferences/IcfgPreferenceInitializer.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+import de.uni_freiburg.informatik.ultimate.core.lib.preferences.UltimatePreferenceInitializer;
+import de.uni_freiburg.informatik.ultimate.core.model.preferences.IPreferenceProvider;
+import de.uni_freiburg.informatik.ultimate.core.model.preferences.PreferenceType;
+import de.uni_freiburg.informatik.ultimate.core.model.preferences.UltimatePreferenceItem;
+import de.uni_freiburg.informatik.ultimate.core.model.services.IUltimateServiceProvider;
+import de.uni_freiburg.informatik.ultimate.lib.modelcheckerutils.boogie.Statements2TransFormula;
+import de.uni_freiburg.informatik.ultimate.lib.smtlibutils.solverbuilder.SolverBuilder.SolverMode;
+import de.uni_freiburg.informatik.ultimate.plugins.generator.icfgbuilder.Activator;
+
+public class IcfgPreferenceInitializer extends UltimatePreferenceInitializer {
```
Comment (schuessf) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/preferences/IcfgPreferenceInitializer.java:42]: Please make sure that these settings are actually the same as for `RCFGBuilder` (this was at least not the case in the past for `CodeBlockSize`).

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/WeakestPrecondition.java [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+import de.uni_freiburg.informatik.ultimate.boogie.ast.IdentifierExpression;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.IfThenElseExpression;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.Procedure;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.QuantifierExpression;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.UnaryExpression;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.UnaryExpression.Operator;
+import de.uni_freiburg.informatik.ultimate.boogie.ast.VarList;
+import de.uni_freiburg.informatik.ultimate.core.model.models.ModelUtils;
+
+public class WeakestPrecondition extends BoogieTransformer {
```
Comment (schuessf) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/WeakestPrecondition.java:50]: I suppose this class (along with some others) was simply copied from `RCFGBuilder` without any change? I think we should rather have another plugin `Library-CFG` with such classes, s.t. `RCFGBuilder` could be safely deleted.

File: trunk/examples/settings/automizer/ForwardPredicates_IcfgBuilder.epf [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+/instance/de.uni_freiburg.informatik.ultimate.plugins.generator.traceabstraction/Compute\ Hoare\ Annotation\ of\ negated\ interpolant\ automaton,\ abstraction\ and\ CFG=true
+@UltimateGui=0.0.1
+#Sat Mar 15 16:00:32 CET 2014
+file_export_version=3.0
+@de.uni_freiburg.informatik.ultimate.core=0.0.1
+/instance/de.uni_freiburg.informatik.ultimate.core/Generate\ benchmark\ results=false
+#Tue Mar 19 19:15:08 CET 2024
+file_export_version=3.0
+\!/instance/de.uni_freiburg.informatik.ultimate.boogie.preprocessor=
+/instance/de.uni_freiburg.informatik.ultimate.boogie.preprocessor/Replace\ while\ statements\ and\ if-then-else\ statements=false
```
Comment (schuessf) [inline trunk/examples/settings/automizer/ForwardPredicates_IcfgBuilder.epf]: This should be our new default now, maybe you can simply change it in `BoogiePreprocessor`.

File: trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/LargeBlockEncoding.java
Comment (schuessf) [inline trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/LargeBlockEncoding.java:1]: Why was this class added to `RCFGBuilder`?

**Review by schuessf — COMMENTED** [review]
Thanks for your work! I did not have a detailed look yet, but here are already some more general comments.

File: trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/LargeBlockEncoding.java
Comment (maul-esel) [reply trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/LargeBlockEncoding.java:1]: I extracted this class from `CfgBuilder` (where it used to be a nested class) so it can be used by both implementations.

File: trunk/examples/Backtranslation/regression/bpl/Pointer.errorpath (hunk 3 above)
Comment (Heizmann) [reply trunk/examples/Backtranslation/regression/bpl/Pointer.errorpath]: Could you fix this?

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/preferences/IcfgPreferenceInitializer.java (hunk 4 above)
Comment (Heizmann) [reply trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/preferences/IcfgPreferenceInitializer.java:42]: Could you fix this?

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/WeakestPrecondition.java (hunk 5 above)
Comment (Heizmann) [reply trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/WeakestPrecondition.java:50]: Plan is to move the classes to from `RCFGBuilder` to `IcfgBuilder` if we want to get rid of the `RCFGBuilder`.

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/preferences/IcfgPreferenceInitializer.java (hunk 4 above)
Comment (schuessf) [reply trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/preferences/IcfgPreferenceInitializer.java:42]: I am not sure, if there is a problem in the current version. I just remember that there was some issue in the past. So this comment is more of a reminder 😉

File: trunk/examples/Backtranslation/regression/bpl/Pointer.errorpath (hunk 3 above)
Comment (schuessf) [reply trunk/examples/Backtranslation/regression/bpl/Pointer.errorpath]: I was just wondering why you changed already this path and not the others.

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/WeakestPrecondition.java (hunk 5 above)
Comment (schuessf) [reply trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/WeakestPrecondition.java:50]: I see, this class is really just used in `IcfgBuilder` / `RCFGBuilder`, so it makes sense to keep it there. But there are some other classes (e.g. `TransFormulaAdder`) that are also used in other Plugins. So it would make sense to move them to some library, also to reduce the dependencies to `IcfgBuilder` / `RCFGBuilder`.

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/preferences/IcfgPreferenceInitializer.java (hunk 4 above)
Comment (schuessf) [reply trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/preferences/IcfgPreferenceInitializer.java:42]: I just checked, the settings are synced right now.

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/WeakestPrecondition.java (hunk 5 above)
Comment (schuessf) [reply trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/WeakestPrecondition.java:50]: Okay, for now we don't create such a library, as this PR should be merged asap

**Comment by schuessf** [comment]
I just synced the `CfgBuilder` classes from `RCFGBuilder` and the new `IcfgBuilder`. To inspect the differences, you can take a look a the [diff](https://github.com/user-attachments/files/17678265/cfgbuilder.patch).

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java [hunk 7]
```diff
@@ -0,0 +1,1627 @@
...
+/**
+ * This class generates a recursive control flow graph (in the style of POPL'10 - Heizmann, Hoenicke, Podelski - Nested
+ * Interpolants) from an Boogie AST which contains only unstructured Code (i.e., no while (and others??) statements).
+ * The input for this classs has to be unstructured Boogie Code (e.g., no while loops) for example the output of the
+ * BoogiePreprocessor.
+ *
+ * @author heizmann@informatik.uni-freiburg.de
+ */
+
+// TODO How to give every location the right line number
```
Comment (schuessf) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java]: Is this TODO fixed?

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java [hunk 8]
```diff
@@ -0,0 +1,1627 @@
...
+			} else {
+				final VariableLHS varlhs = (VariableLHS) lhs;
+				expr = new IdentifierExpression(lhs.getLocation(), lhs.getType(), varlhs.getIdentifier(),
+						varlhs.getDeclarationInformation());
+			}
+			return expr;
+		}
+
+		private IIcfgElement buildAssuAssiHavoc(final IIcfgElement currentElement, Statement st) {
+			// FIXME Matthias 2024-09-22: This code is extremely problematic because it
```
Comment (schuessf) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java]: @Heizmann Are you planning to change this? Why isn't this necessary in the old `RCFGBuilder`?

**Comment by schuessf** [comment]
@Heizmann I just ran the nightly build. Unfortunately, currently not only the backtranslation tests fail (which is expected and could be quite easily fixed), but also [generic regression tests](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate%20Nightly/job/PR-690/1/testReport/de.uni_freiburg.informatik.ultimate.regressiontest.generic/).
This includes some LTL tests (e.g. [here](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate%20Nightly/job/PR-690/1/testReport/de.uni_freiburg.informatik.ultimate.regressiontest.generic/LTLRegressionTestSuite/I_LTL_regression_c_stepwise_baseline_coolant_basis_4_neg_c_S_LTL_regression_c_LTLAutomizerC_epf_T_LTL_regression_c_LTLAutomizerC_xml/)), where we don't find any specification, some referee tests (e.g. [here](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate%20Nightly/job/PR-690/1/testReport/de.uni_freiburg.informatik.ultimate.regressiontest.generic/RegressionTestSuite/I_programs_toy_InvariantChecking_regression_bpl_OnlyRequires01_bpl_S_programs_toy_InvariantChecking_regression_bpl_RefereeBpl_epf_T_programs_toy_InvariantChecking_regression_bpl_RefereeBpl_xml/)), where we claim that there is a cycle now and some concurrency tests (e.g. [here](https://jenkins.sopranium.de/job/Ultimate/job/Ultimate%20Nightly/job/PR-690/1/testReport/de.uni_freiburg.informatik.ultimate.regressiontest.generic/RegressionTestSuite/I_concurrent_bpl_regression_concurrent_increment_bpl_S_concurrent_bpl_regression_ReachSafety_32bit_GemCutter_epf_T_concurrent_bpl_regression_ReachSafety_xml/)) that now have a timeout, maybe because of the `assume true` statements that are not removed (see [above](https://github.com/ultimate-pa/ultimate/pull/690#issuecomment-2455268452))

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java [hunk 9]
```diff
@@ -0,0 +1,1627 @@
...
+		// implementation
+		final BoogieIcfgContainer icfg = mIcfg;
+		for (final String procName : mBoogieDeclarations.getProcImplementation().keySet()) {
+			final Body body = mBoogieDeclarations.getProcImplementation().get(procName).getBody();
+			final Statement firstStatement;
+			if (body.getBlock().length == 0) {
+				firstStatement =
+						new ReturnStatement(mBoogieDeclarations.getProcImplementation().get(procName).getLocation());
+			} else {
+				firstStatement = body.getBlock()[0];
```
Comment (schuessf) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:269]: In the `RCFGBuilder` `firstStatement` is always set to `body.getBlock()[0]`, why is this different here?

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java [hunk 10]
```diff
@@ -0,0 +1,1627 @@
...
+			// mlocSuffix = 0;
+			mCurrent = exitNode;
+
+			for (final EnsuresSpecification spec : ensures) {
+				final AssumeStatement st = new AssumeStatement(spec.getLocation(), spec.getFormula());
+				ModelUtils.copyAnnotations(spec, st);
+				mIcfgBacktranslator.putAux(st, new BoogieASTNode[] { spec });
+				mCurrent = buildAssuAssiHavoc((IIcfgElement) mCurrent, st);
+			}
+			final BoogieIcfgLocation finalNode = mIcfg.mFinalNode.get(mCurrentProcedureName);
```
Comment (schuessf) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:1423]: In the `RCFGBuilder` the exit nodes are used, while here the final nodes are used. Is this change intended and what are the implications?

**Comment by schuessf** [comment]
I just looked into the open backtranslation issues, the created CFG for loops (especially when translated from C) looks slightly different.
Consider the following C program
```c
while (x > 0) x--
```
This is roughly translated to the following Boogie program
```c
while (true) {
  label: if (!(x > 0)) {
    break;
  }
  x := x - 1;
}
```
In the old CFG, we only had only two states and three edges (with `SequenceOfStatements`):
* a loop in the initial location with the body and the negated if-condition (`x > 0`)
* an edge from the initial location to the exit location with the if-condition (`!(x > 0)`)
* an edge from the initial location to the exit location with the negated while-condition (`false`)
grafik


With the same settings, the new CFG has 5 locations and the following 6 edges:
* one with the while condition (`true`)
* an edge from the initial location to the exit location with the negated while-condition (`false`)
* another edge with `true` after the label
* an edge with the negated if-condition (`x > 0`)
* an edge to the exit location with the if-condition (`!(x > 0)`)
* an edge with the body back to the initial location
grafik

This means that in the old CFG we had a stronger block encoding for `SequenceOfStatements`. With other settings the CFG should be more similar. I am not sure if this behaviour is desired.

**Comment by Heizmann** [comment]
Some info (not enough time to think about your post): The new IcfgBuilder alredy makes sure that labels get their own IcfgLocation.

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java (hunk 10 above)
Comment (NiklasKult) [reply trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:1423]: The RCFGBuilder builds from the top, while the IcfgBuilder builds from the bottom. Because of that the ensures is built from the exit node instead of the final node.
The resulting CFG should be identical here.

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java (hunk 9 above)
Comment (NiklasKult) [reply trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:269]: For the <code class="notranslate">RCFGBuilder</code> <code class="notranslate">body.getBlock()</code> could never be empty, because the <code class="notranslate">BoogiePreprocessor</code> would insert a return statement in that case. With the setting that disables the <code class="notranslate">UnstructureCode</code> class from the <code class="notranslate">BoogiePreprocessor</code> <code class="notranslate">body.getBlock()</code> can now be empty.

Comment (schuessf) [reply trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:269]: But does this belong to the `CfgBuilder`? Wouldn't it be better to move it into its own Preprocessor that always runs independant of `UnstructureCode`?

Comment (schuessf) [reply trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:269]: Or do we even need this statement here? Is it possible to create a location without a statement?

File: trunk/source/IcfgBuilder/.classpath [hunk 11]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<classpath>
+	<classpathentry kind="con" path="org.eclipse.jdt.launching.JRE_CONTAINER/org.eclipse.jdt.internal.debug.ui.launcher.StandardVMType/JavaSE-21"/>
+	<classpathentry exported="true" kind="con" path="org.eclipse.pde.core.requiredPlugins"/>
+	<classpathentry kind="src" path="src"/>
+	<classpathentry combineaccessrules="false" kind="src" path="/BoogiePrinter"/>
```
Comment (bahnwaerter) [inline trunk/source/IcfgBuilder/.classpath:6]: Why is this access rule necessary? Does a utility need to be accessed internally here?
If so, then the API and implementation of the required software component should be restructured to avoid such hacks and instead access public interfaces cleanly via the `MANIFEST.MF`.

File: trunk/source/IcfgBuilder/LicenseTemplate.md [hunk 12]
```diff
@@ -0,0 +1,23 @@
+Copyright (C) @{daterange} @{author:r}
+
+This file is part of the ULTIMATE RCFGBuilder plug-in.
```
Comment (bahnwaerter) [inline trunk/source/IcfgBuilder/LicenseTemplate.md]: Shouldn't the objective be called ICFGBuilder instead of RCFGBuilder in the whole license template file?

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/IcfgBacktranslator.java [hunk 13]
```diff
@@ -0,0 +1,471 @@
...
+
+	@Override
+	public IBacktranslatedCFG<String, BoogieASTNode>
+			translateCFG(final IBacktranslatedCFG<IcfgLocation, IIcfgTransition<IcfgLocation>> cfg) {
+		final IBacktranslatedCFG<String, BoogieASTNode> translatedCfg =
+				translateCFG(cfg, (a, b, c) -> translateCFGEdge(a, (IIcfgTransition<IcfgLocation>) b, c));
+		// mLogger.info(getClass().getSimpleName());
+		// printHondas(cfg, mLogger::info);
+		// printCFG(cfg, mLogger::info);
+		// mLogger.info("######## END "+getClass().getSimpleName());
```
Comment (bahnwaerter) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/IcfgBacktranslator.java]: Remove unused debugging code or implement it correctly when debug mode is enabled.

**Review by bahnwaerter — COMMENTED** [review]
The changes look good so far. The branch has already been correctly adapted to our Java 21 and Eclipse migration and can therefore be merged without any issues.

However, there are still a few small things that need to be improved. For example, in the comments and the license agreement in the headers of ICFG source code files, the word _RCFGBuilder_ should be replaced consistently with _ICFGBuilder_.

File: trunk/source/IcfgBuilder/.classpath (hunk 11 above)
Comment (schuessf) [reply trunk/source/IcfgBuilder/.classpath:6]: I don't know, this is the same as in `RCFGBuilder`, so I will just keep this for now, we can change this later (after SV-COMP 😉)

File: trunk/source/AbstractInterpretationV2/src/de/uni_freiburg/informatik/ultimate/plugins/analysis/abstractinterpretationv2/AbstractInterpretation.java [hunk 14]
```diff
@@ -77,10 +77,12 @@ public void setInputDefinition(final ModelType graphType) {
 		final String creator = graphType.getCreator();
 		switch (creator) {
 		case "de.uni_freiburg.informatik.ultimate.plugins.generator.rcfgbuilder":
+		case "de.uni_freiburg.informatik.ultimate.plugins.generator.icfgbuilder":
 			mObserver = new ArrayList<>();
 			mObserver.add(new AbstractInterpretationRcfgObserver(mServices));
 			break;
 		default:
+			mLogger.info("Ignoring creator " + creator);
```
Comment (maul-esel) [inline trunk/source/AbstractInterpretationV2/src/de/uni_freiburg/informatik/ultimate/plugins/analysis/abstractinterpretationv2/AbstractInterpretation.java]: Nitpick: Make this a warning

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/IcfgBacktranslator.java [hunk 15]
```diff
@@ -0,0 +1,467 @@
...
+		} else {
+			throw new UnsupportedOperationException("Unsupported CodeBlock: " + cb.getClass().getCanonicalName());
+		}
+		trace.add(ateBuilder.build());
+	}
+
+	@Override
+	public IProgramExecution<BoogieASTNode, Expression>
+
+			translateProgramExecution(final IProgramExecution<IIcfgTransition<IcfgLocation>, Term> programExecution) {
```
Comment (maul-esel) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/IcfgBacktranslator.java:286]: nitpick: Fix the formatting

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/IcfgBacktranslator.java [hunk 16]
```diff
@@ -0,0 +1,467 @@
...
+		if (!(programExecution instanceof IcfgProgramExecution)) {
+			throw new IllegalArgumentException();
```
Comment (maul-esel) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/IcfgBacktranslator.java]: Nitpick: add an error message

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java [hunk 17]
```diff
@@ -0,0 +1,1617 @@
...
+				Logics.valueOf(prefs.getString(IcfgPreferenceInitializer.LABEL_EXT_SOLVER_LOGIC));
+		final SolverSettings solverSettings =
+				SolverBuilder.constructSolverSettings().setUseFakeIncrementalScript(fakeNonIncrementalScript)
+						.setDumpSmtScriptToFile(dumpSmtScriptToFile, pathOfDumpedScript, filename, compressSmtScript)
+						.setDumpUnsatCoreTrackBenchmark(dumpUnsatCoreTrackBenchmark)
+						.setDumpMainTrackBenchmark(dumpMainTrackBenchmark)
+						.setUseExternalSolver(true, commandExternalSolver, logicForExternalSolver)
+						.setSolverMode(solverMode).setAdditionalOptions(additionalSmtOptions);
+
+		return SolverBuilder.buildAndInitializeSolver(services, solverSettings, "CfgBuilderScript");
```
Comment (maul-esel) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java:383]: In the future, we could wrap these settings in a suitable class rather than accessing the `IPreferenceProvider` directly here.
This should not be an obstacle for the merge now though.

File: trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java [hunk 18]
```diff
@@ -0,0 +1,1617 @@
...
+			ModelUtils.copyAnnotations(st, thenCondition);
+			ModelUtils.copyAnnotations(st, elseCondition);
+			buildBranching(thenCondition, thenPart, elseCondition, elsePart, srcLoc);
+			// remove end node for LoopFreeBlock and SequenceOfStatements, if it only has one incoming edge and one
+			// outgoing edge
+			if ((mCodeBlockSize == CodeBlockSize.LoopFreeBlock || mCodeBlockSize == CodeBlockSize.SequenceOfStatements)
+					&& endLoc.getIncomingEdges().size() == 1
+					&& endLoc.getOutgoingEdges().size() == 1
+					&& !mConditionalStarts.contains(endLoc)
+					&& !mLabel2LocNodes.containsValue(endLoc)) {
```
Comment (maul-esel) [inline trunk/source/IcfgBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/icfgbuilder/cfg/CfgBuilder.java]: A complex condition like that could be encapsulated in a method with an intuitive name.

File: trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/BoogieIcfgContainer.java [hunk 19]
```diff
@@ -89,7 +89,7 @@ public class BoogieIcfgContainer extends ModernAnnotations implements IIcfg<Boog
 	 * Note: This map is only used during construction!
 	 *
 	 */
-	final Map<String, BoogieIcfgLocation> mFinalNode;
+	public final Map<String, BoogieIcfgLocation> mFinalNode;
```
Comment (maul-esel) [inline trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/BoogieIcfgContainer.java:92]: Is it necessary to make this public?

**Review by maul-esel — APPROVED** [review]
Thanks @NiklasKult, @Heizmann and @schuessf for working on this!

I had a look at the changes and did not find any major issues. Due to the time constraints, I was not able to review in-depth, but since others have done so, and our latest evaluations have shown no major issues (only a slight increase in the number of locations) I have no objection to merging :+1:

**Comment by maul-esel** [comment]
I also had a quick look at the RCFGBuilder/ICFGBuilder diff by @schuessf and did not find any issues there either.

**Comment by schuessf** [comment]
I will try to address these minor issues and merge this later.

File: trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/BoogieIcfgContainer.java (hunk 19 above)
Comment (schuessf) [reply trunk/source/RCFGBuilder/src/de/uni_freiburg/informatik/ultimate/plugins/generator/rcfgbuilder/cfg/BoogieIcfgContainer.java:92]: I think so, as this is also used by the IcfgBuilder plugin. Of course, it would be more optimal to move it to some library, but we can do this, if we remove RCFGBuilder in the future.

**Review by bahnwaerter — APPROVED** [review]
